### PR TITLE
[draft] testing for moving precision plugin as property of TrainingTypePlugin

### LINKED ISSUE #7324: Precision Plugins should be part of Training Type Plugins
## 🚀 Feature

Precision Plugin should be a property of the training type plugin, not the accelerator. these can overlap, and interleaving sequences of operations across the two in the accelerator can be challenging to get fully right.

cc @justusschock @awaelchli @kaushikb11 @tchaton 

todo:  fill out in more detail

### Motivation

This would make management of things like this easier:
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7271
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7217

### PR DESCRIPTION
**Pull request by shuyingsunshine21**
## What does this PR do?

Fixes #7324

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -201,7 +199,7 @@ def training_step(
         """
         step_kwargs = self.to_device(step_kwargs)
 
-        with self.precision_plugin.train_step_context(), self.training_type_plugin.train_step_context():
+        with self.training_type_plugin.precision_plugin.train_step_context(), self.training_type_plugin.train_step_context():
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py]: We basically have two different contexts here: One for forward + backward (current train_step_context) and one for forward only.

I think we could change it down to these two as well in this PR. From my perspective there is no difference between forward in training and forward in val/test/predict so we should be able to combine them.

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -52,21 +52,16 @@ class Accelerator:
...
-        precision_plugin: PrecisionPlugin,
         training_type_plugin: TrainingTypePlugin,
     ) -> None:
         """
         Args:
             precision_plugin: the plugin to handle precision-specific parts
             training_type_plugin: the plugin to handle different training routines
         """
-        self.precision_plugin = precision_plugin
+        self.precision_plugin = training_type_plugin.precision_plugin
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:62]: do we even need a fixed reference here then? I think replacing it with a property might be better since that does not increase the reference counter of the object

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -291,63 +289,16 @@ def backward(
...
-        if make_optimizer_step:
-            self.run_optimizer_step(optimizer, opt_idx, lambda_closure, **kwargs)
-        self.precision_plugin.post_optimizer_step(optimizer, opt_idx)
-        self.training_type_plugin.post_optimizer_step(optimizer, opt_idx, **kwargs)
-
-    def run_optimizer_step(
-        self, optimizer: Optimizer, optimizer_idx: int, lambda_closure: Callable, **kwargs: Any
-    ) -> None:
-        self.training_type_plugin.optimizer_step(optimizer, lambda_closure=lambda_closure, **kwargs)
+        self.training_type_plugin.optimizer_step(optimizer, opt_idx, lambda_closure, **kwargs)
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:299]: The only downside here is that the training type plugin must make sure to call the precision plugin here. This is something the user has to make sure when implementing his own training type plugin

File: pytorch_lightning/plugins/training_type/ddp2.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def global_rank(self) -> int:
     def world_size(self) -> int:
         return self.num_nodes
 
-    def setup(self, model):
+    def setup_model(self, model):
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp2.py:31]: I'd rather not rename these, since they aren't meant to setup only the model. They can actually setup arbitrary things

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 5]
```diff
@@ -54,9 +54,9 @@ def backward(
         self,
         model: 'pl.LightningModule',
         closure_loss: Tensor,
+        should_accumulate: bool,
```
Comment (justusschock) [inline pytorch_lightning/plugins/precision/precision_plugin.py:57]: Not sure, but i think the order should be preserved here to avoid BC for positional args

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 6]
```diff
@@ -132,7 +132,7 @@ def distributed_sampler_kwargs(self):
     def _is_single_process_single_device(self):
         return True
 
-    def setup(self, model):
+    def setup_model(self, model):
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:135]: same as above

File: pytorch_lightning/plugins/training_type/dp.py [hunk 7]
```diff
@@ -47,7 +47,7 @@ def node_rank(self) -> int:
     def world_size(self) -> int:
         return 1
 
-    def setup(self, model):
+    def setup_model(self, model):
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/dp.py:50]: same as above

File: pytorch_lightning/plugins/training_type/sharded.py [hunk 8]
```diff
@@ -54,7 +67,7 @@ def _reinit_optimizers_with_oss(self):
                 optim_class = type(optimizer)
                 zero_optimizer = OSS(params=optimizer.param_groups, optim=optim_class, **optimizer.defaults)
                 if _FAIRSCALE_OSS_FP16_BROADCAST_AVAILABLE:
-                    is_fp16 = self.lightning_module.trainer.precision == 16
+                    is_fp16 = self.lightning_module.trainer.precision == "mixed"
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/sharded.py:68]: wasn't that properly handled before?

File: pytorch_lightning/plugins/training_type/single_device.py [hunk 9]
```diff
@@ -66,9 +66,8 @@ def model_to_device(self) -> None:
 
         self._model.to(self.root_device)
 
-    def setup(self, model: torch.nn.Module) -> torch.nn.Module:
+    def setup_model(self, model: 'pl.LightningModule') -> None:
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/single_device.py:69]: same as above

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 10]
```diff
@@ -111,9 +111,8 @@ def pre_dispatch(self):
         if self.debug:
             os.environ["PT_XLA_DEBUG"] = str(1)
 
-    def setup(self, model: Module) -> Module:
+    def setup_model(self, model: 'pl.LightningModule') -> None:
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:114]: same as above

File: pytorch_lightning/plugins/training_type/horovod.py [hunk 11]
```diff
@@ -55,7 +55,7 @@ def distributed_sampler_kwargs(self):
         distributed_sampler_kwargs = dict(num_replicas=self.world_size, rank=self.global_rank)
         return distributed_sampler_kwargs
 
-    def setup(self, model):
+    def setup_model(self, model):
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/horovod.py:58]: same as above

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 12]
```diff
@@ -41,6 +57,84 @@ def __init__(self) -> None:
...
+        self.optimizers: List = []
+        self.lr_schedulers: List = []
+        self.optimizer_frequencies: List = []
+
+    @property
+    def precision_plugin(self) -> Optional[PrecisionPlugin]:
+        return self._precision_plugin
+
+    @precision_plugin.setter
+    def precision_plugin(self, args: Dict[str, Any]) -> None:
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:67]: tbh, I don't like this. This should live in the acceleator connector. You perform the usual checks there and check if this class implements some other logic, but ideally they'd be determined beforehand (maybe as a staticmethod since you don't seem to really need the sate for selection)

**Review by justusschock — COMMENTED** [review]
Some initial comments.

All in all I am not sure, if I really like this. I see the convenience having this, but i also think this leads in the wrong direction (back to monolithic class structures) and entangles a lot of the logic we want to be separated from other parts again.

This includes (but isn't limited to)
* no more strong separation of training and precision plugin (yes I know they're still separate classes but you don't need a precision plugin anymore; you can do all in the training type plugin (and out of convenience and lazyness people will do that))
* Users have to care about both plugins when only implementing the training type (at least need to call hooks from precision)
* The accelerator as an orchestration class becomes basically useless.

So in total this feels to me like we are heading back to fixed combinations of training and precision plugins (partly even incorporating the accelerator) with no clear responsibility anymore.

**Comment by shuyingsunshine21** [comment]
Thanks @justusschock  for the comments!

I am also thinking whether going back to one class like `TrainingStrategyPlugin` to control everything is a bad thing or not given that interleaving calls across TrainingType and Precision handled by accelerator might be problematic. This flow might cause some combination of TrainingType and Precision do not work as expected.

>  no more strong separation of training and precision plugin (yes I know they're still separate classes but you don't need a precision plugin anymore; you can do all in the training type plugin (and out of convenience and lazyness people will do that))

I also feel that it is a bit vague to separate the current training type plugin and precision plugin. One can actually do everything in training type plugin and has some dummy precision plugin even for the current way. If we would like to have clear separation, what would be a good interface to ensure that.

> Users have to care about both plugins when only implementing the training type (at least need to call hooks from precision)

I am thinking this is actually intended, as this is to avoid implementing one of them without caring logics of the other, causing unexpected behavior. Current way, user need to care about this in accelerator level though.

File: pytorch_lightning/plugins/training_type/sharded.py (hunk 8 above)
Comment (shuyingsunshine21) [reply pytorch_lightning/plugins/training_type/sharded.py:68]: yeah, unfortunately, will separate this small fix.
